python/src/scipp/io/hdf5.py

```diff
@@ -0,0 +1,102 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+# @file
+# @author Simon Heybrock
+
+
+def _unit_lut():
+    from .._scipp.core import units as u
+    units = [u.one, u.m, u.kg, u.us, u.angstrom]
+    names = [str(unit) for unit in units]
+    return dict(zip(names, units))
+
+
+class HDF5IO:
+    _units = _unit_lut()
+
+    def _read_unit(self, dset):
+        return self._units[dset.attrs['unit']]
+
+    def _write_variable(self, group, var, name):
+        from .._scipp import core as sc
+        if var.dtype in [sc.dtype.string, sc.dtype.PyObject]:
+            print(f'Writing with dtype={var.dtype} not implemented, skipping.')
+            return
+        group = group.create_group(name)
+        dset = group.create_dataset('values', data=var.values)
+        dset.attrs['dims'] = [str(dim) for dim in var.dims]
+        dset.attrs['dtype'] = str(var.dtype)
+        dset.attrs['unit'] = str(var.unit)
+        if var.variances is not None:
+            variances = group.create_dataset('variances', data=var.variances)
+            dset.attrs['variances'] = variances.ref
+
+    def _read_variable(self, group):
+        from .._scipp import core as sc
+        import numpy as np
+        values = group['values']
+        dtype = values.attrs['dtype']
+        contents = dict()
+        # For some reason we cannot pass the dataset directly to scipp, and
+        # we also need to manually override the dtype. Is h5py using a
+        # different float type?
+        contents['values'] = np.array(values, dtype=dtype, copy=False)
+        contents['dims'] = values.attrs['dims']
+        contents['unit'] = self._read_unit(values)
+        if 'variances' in group:
+            contents['variances'] = np.array(group['variances'],
+                                             dtype=dtype,
+                                             copy=False)
+        return sc.Variable(**contents)
+
+    def _write_data_array(self, group, data):
+        import scipp as sc
+        group.attrs['scipp-version'] = sc.__version__
+        group.attrs['type'] = 'DataArray'
+        group.attrs['name'] = data.name
+        if data.data is None:
+            raise RuntimeError(
+                "Writing realigned data is not implemented yet.")
+        self._write_variable(group, var=data.data, name='data')
+        coords = group.create_group('coords')
+        for name in data.coords:
+            self._write_variable(group=coords,
+                                 var=data.coords[name],
+                                 name=str(name))
+        masks = group.create_group('masks')
+        for name in data.masks:
+            self._write_variable(group=masks, var=data.masks[name], name=name)
+        attrs = group.create_group('attrs')
+        for name in data.attrs:
+            self._write_variable(group=attrs, var=data.attrs[name], name=name)
```

Comment:
This code is almost the same for `coords`, `masks` and `attrs`. Any way you can refactor to limit duplication?